From: Michael Froman <mfroman@mozilla.com>
Date: Thu, 6 Aug 2026 13:25:11 +0000
Subject: Bug 2053579 - avoid DescendingSeqNumComp<uint16_t> as comparator in
 std::set. a=diannaS DONTBUILD
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

This is a combined cherry-pick of upstream libwebrtc bf6f744d19, 5ecea90550,
6395da1ed7.

Upstream commit: https://webrtc.googlesource.com/src/+/bf6f744d19a84a7d0c3f5acbc844123a8197165a
       Check RTP sequence number order when assembling H264 frame in PacketBuffer

       PacketBuffer::buffer_ slots can be re-used by packets with with
       different sequence numbers when their difference is multiple of
       buffer_.size()

       Non-H264 code path validates RTP sequence numbers continuity in
       `PotentialNewFrame` helper, H264 code path missed such check as the new
       test demonstrate.

       Bug: chromium:537233963
       Change-Id: I6af0f52bfcb3a4437450dffcece85bee7ce172cd
       Reviewed-on: https://webrtc-review.googlesource.com/c/src/+/490840
       Reviewed-by: Åsa Persson <asapersson@webrtc.org>
       Commit-Queue: Danil Chapovalov <danilchap@webrtc.org>
       Cr-Commit-Position: refs/heads/main@{#48222}

Upstream commit: https://webrtc.googlesource.com/src/+/5ecea90550816628fa124e7cf55f4cdbc5aa4a7e
       In PacketBuffer use helper to find packet slot in the buffer

       This is a preparation for using unwrapped int64_t instead of wrapped
       uint16_t RTP sequence number in PacketBuffer. Using modular works when
       sequence number is unsigned but produces incorrect result when signed
       sequence number happens to be negative.

       Bug: chromium:537233963
       Change-Id: I515cb7441d052a2823eee4891bb4e17b254d602f
       Reviewed-on: https://webrtc-review.googlesource.com/c/src/+/491920
       Commit-Queue: Danil Chapovalov <danilchap@webrtc.org>
       Reviewed-by: Ilya Nikolaevskiy <ilnik@webrtc.org>
       Cr-Commit-Position: refs/heads/main@{#48231}

Upstream commit: https://webrtc.googlesource.com/src/+/6395da1ed7da9f3a8d97475c4e8db722d9415e6f
       Use unwrappred RTP sequence numbers in PacketBuffer

       Thus avoid brittle DescendingSeqNumComp<uint16_t> as comparator in set.
       Such comparator lacks some properties like transitivity making it
       dangerous to use with the std::set

       Bug: chromium:537233963
       Change-Id: I400b423ab3f0d557219d459752e00d211fd97450
       Reviewed-on: https://webrtc-review.googlesource.com/c/src/+/491041
       Commit-Queue: Danil Chapovalov <danilchap@webrtc.org>
       Reviewed-by: Ilya Nikolaevskiy <ilnik@webrtc.org>
       Cr-Commit-Position: refs/heads/main@{#48232}

Differential Revision: https://phabricator.services.mozilla.com/D316661
Mercurial Revision: https://hg.mozilla.org/mozilla-central/rev/8a74dd6c25d7
---
 api/video/rtp_video_frame_assembler.cc        | 27 +++---
 modules/video_coding/packet_buffer.cc         | 86 +++++++++----------
 modules/video_coding/packet_buffer.h          | 32 ++++---
 .../video_coding/packet_buffer_unittest.cc    | 10 ++-
 video/rtp_video_stream_receiver2.cc           | 31 ++++---
 video/rtp_video_stream_receiver2.h            |  2 +-
 6 files changed, 104 insertions(+), 84 deletions(-)

diff --git a/api/video/rtp_video_frame_assembler.cc b/api/video/rtp_video_frame_assembler.cc
index 8c0316232b..c02c7a218b 100644
--- a/api/video/rtp_video_frame_assembler.cc
+++ b/api/video/rtp_video_frame_assembler.cc
@@ -91,12 +91,12 @@ class RtpVideoFrameAssembler::Impl {
   RtpFrameVector AssembleFrames(
       video_coding::PacketBuffer::InsertResult insert_result);
   FrameVector FindReferences(RtpFrameVector frames);
-  FrameVector UpdateWithPadding(uint16_t seq_num);
+  FrameVector UpdateWithPadding(int64_t seq_num);
   bool ParseDependenciesDescriptorExtension(const RtpPacketReceived& rtp_packet,
                                             RTPVideoHeader& video_header);
   bool ParseGenericDescriptorExtension(const RtpPacketReceived& rtp_packet,
                                        RTPVideoHeader& video_header);
-  void ClearOldData(uint16_t incoming_seq_num);
+  void ClearOldData(int64_t incoming_seq_num);
 
   std::unique_ptr<FrameDependencyStructure> video_structure_;
   SeqNumUnwrapper<uint16_t> rtp_sequence_number_unwrapper_;
@@ -114,9 +114,11 @@ RtpVideoFrameAssembler::Impl::Impl(
 
 RtpVideoFrameAssembler::FrameVector RtpVideoFrameAssembler::Impl::InsertPacket(
     const RtpPacketReceived& rtp_packet) {
+  int64_t unwrapped_sequence_number =
+      rtp_sequence_number_unwrapper_.Unwrap(rtp_packet.SequenceNumber());
   if (rtp_packet.payload_size() == 0) {
-    ClearOldData(rtp_packet.SequenceNumber());
-    return UpdateWithPadding(rtp_packet.SequenceNumber());
+    ClearOldData(unwrapped_sequence_number);
+    return UpdateWithPadding(unwrapped_sequence_number);
   }
 
   std::optional<VideoRtpDepacketizer::ParsedRtpPayload> parsed_payload =
@@ -146,16 +148,16 @@ RtpVideoFrameAssembler::FrameVector RtpVideoFrameAssembler::Impl::InsertPacket(
       parsed_payload->video_header);
   packet->video_payload = std::move(parsed_payload->video_payload);
 
-  ClearOldData(rtp_packet.SequenceNumber());
+  ClearOldData(unwrapped_sequence_number);
   return FindReferences(
       AssembleFrames(packet_buffer_.InsertPacket(std::move(packet))));
 }
 
-void RtpVideoFrameAssembler::Impl::ClearOldData(uint16_t incoming_seq_num) {
+void RtpVideoFrameAssembler::Impl::ClearOldData(int64_t incoming_seq_num) {
   constexpr uint16_t kOldSeqNumThreshold = 2000;
-  uint16_t old_seq_num = incoming_seq_num - kOldSeqNumThreshold;
+  int64_t old_seq_num = incoming_seq_num - kOldSeqNumThreshold;
   packet_buffer_.ClearTo(old_seq_num);
-  reference_finder_.ClearTo(old_seq_num);
+  reference_finder_.ClearTo(static_cast<uint16_t>(old_seq_num));
 }
 
 RtpVideoFrameAssembler::Impl::RtpFrameVector
@@ -184,8 +186,8 @@ RtpVideoFrameAssembler::Impl::AssembleFrames(
       // received time from RtpPacketReceived::arrival_time.
       const video_coding::PacketBuffer::Packet& last_packet = *packet;
       result.push_back(std::make_unique<RtpFrameObject>(
-          first_packet->seq_num(),                              //
-          last_packet.seq_num(),                                //
+          static_cast<uint16_t>(first_packet->seq_num()),       //
+          static_cast<uint16_t>(last_packet.seq_num()),         //
           last_packet.marker_bit,                               //
           /*times_nacked=*/0,                                   //
           /*first_packet_received_time=*/Timestamp::Zero(),     //
@@ -224,10 +226,11 @@ RtpVideoFrameAssembler::Impl::FindReferences(RtpFrameVector frames) {
 }
 
 RtpVideoFrameAssembler::FrameVector
-RtpVideoFrameAssembler::Impl::UpdateWithPadding(uint16_t seq_num) {
+RtpVideoFrameAssembler::Impl::UpdateWithPadding(int64_t seq_num) {
   auto res =
       FindReferences(AssembleFrames(packet_buffer_.InsertPadding(seq_num)));
-  auto ref_finder_update = reference_finder_.PaddingReceived(seq_num);
+  auto ref_finder_update =
+      reference_finder_.PaddingReceived(static_cast<uint16_t>(seq_num));
 
   for (std::unique_ptr<RtpFrameObject>& complete_frame : ref_finder_update) {
     uint16_t rtp_seq_num_start = complete_frame->first_seq_num();
diff --git a/modules/video_coding/packet_buffer.cc b/modules/video_coding/packet_buffer.cc
index d4e1d251df..9b8d30f897 100644
--- a/modules/video_coding/packet_buffer.cc
+++ b/modules/video_coding/packet_buffer.cc
@@ -26,8 +26,6 @@
 #include "modules/video_coding/codecs/h264/include/h264_globals.h"
 #include "rtc_base/checks.h"
 #include "rtc_base/logging.h"
-#include "rtc_base/numerics/mod_ops.h"
-#include "rtc_base/numerics/sequence_number_util.h"
 
 namespace webrtc {
 namespace video_coding {
@@ -67,21 +65,20 @@ PacketBuffer::InsertResult PacketBuffer::InsertPacket(
     std::unique_ptr<PacketBuffer::Packet> packet) {
   PacketBuffer::InsertResult result;
 
-  uint16_t seq_num = packet->seq_num();
-  size_t index = seq_num % buffer_.size();
+  const int64_t seq_num = packet->seq_num();
+  size_t index = Index(seq_num);
 
   if (!first_packet_received_) {
     first_seq_num_ = seq_num;
     first_packet_received_ = true;
-  } else if (AheadOf(first_seq_num_, seq_num)) {
+  } else if (first_seq_num_ > seq_num) {
     // If we have explicitly cleared past this packet then it's old,
     // don't insert it, just silently ignore it.
     if (is_cleared_to_first_seq_num_) {
       return result;
     }
 
-    if (ForwardDiff<uint16_t>(first_seq_num_, seq_num) >= max_size_ &&
-        ForwardDiff<uint16_t>(seq_num, first_seq_num_) >= max_size_ / 2) {
+    if (first_seq_num_ - seq_num >= static_cast<int64_t>(max_size_ / 2)) {
       // Large negative jump in rtp sequence number: clear the buffer and treat
       // latest packet as the new first packet.
       Clear();
@@ -93,14 +90,16 @@ PacketBuffer::InsertResult PacketBuffer::InsertPacket(
 
   if (buffer_[index] != nullptr) {
     // Duplicate packet, just delete the payload.
-    if (buffer_[index]->seq_num() == packet->seq_num()) {
+    if (buffer_[index]->seq_num() == seq_num) {
       return result;
     }
 
     // The packet buffer is full, try to expand the buffer.
-    while (ExpandBufferSize() && buffer_[seq_num % buffer_.size()] != nullptr) {
+    // Note that `ExpandBufferSize` changes `buffer_.size()` and thus `Index`
+    // has to be recalculated after it.
+    while (ExpandBufferSize() && buffer_[Index(seq_num)] != nullptr) {
     }
-    index = seq_num % buffer_.size();
+    index = Index(seq_num);
 
     // Packet buffer is still full since we were unable to expand the buffer.
     if (buffer_[index] != nullptr) {
@@ -126,9 +125,9 @@ PacketBuffer::InsertResult PacketBuffer::InsertPacket(
   return result;
 }
 
-uint32_t PacketBuffer::ClearTo(uint16_t seq_num) {
+uint32_t PacketBuffer::ClearTo(int64_t seq_num) {
   // We have already cleared past this sequence number, no need to do anything.
-  if (AheadOf<uint16_t>(first_seq_num_, seq_num)) {
+  if (first_seq_num_ > seq_num) {
     return 0;
   }
 
@@ -140,11 +139,11 @@ uint32_t PacketBuffer::ClearTo(uint16_t seq_num) {
   // iterations to the `size_` of the buffer.
   ++seq_num;
   uint32_t num_cleared_packets = 0;
-  size_t diff = ForwardDiff<uint16_t>(first_seq_num_, seq_num);
-  size_t iterations = std::min(diff, buffer_.size());
+  size_t iterations =
+      std::min<size_t>(seq_num - first_seq_num_, buffer_.size());
   for (size_t i = 0; i < iterations; ++i) {
-    auto& stored = buffer_[first_seq_num_ % buffer_.size()];
-    if (stored != nullptr && AheadOf<uint16_t>(seq_num, stored->seq_num())) {
+    auto& stored = buffer_[Index(first_seq_num_)];
+    if (stored != nullptr && seq_num > stored->seq_num()) {
       ++num_cleared_packets;
       stored = nullptr;
     }
@@ -169,11 +168,11 @@ void PacketBuffer::Clear() {
   ClearInternal();
 }
 
-PacketBuffer::InsertResult PacketBuffer::InsertPadding(uint16_t seq_num) {
+PacketBuffer::InsertResult PacketBuffer::InsertPadding(int64_t seq_num) {
   PacketBuffer::InsertResult result;
   UpdateMissingPackets(seq_num);
   received_padding_.insert(seq_num);
-  result.packets = FindFrames(static_cast<uint16_t>(seq_num + 1));
+  result.packets = FindFrames(seq_num + 1);
   return result;
 }
 
@@ -208,7 +207,7 @@ bool PacketBuffer::ExpandBufferSize() {
   std::vector<std::unique_ptr<Packet>> new_buffer(new_size);
   for (std::unique_ptr<Packet>& entry : buffer_) {
     if (entry != nullptr) {
-      new_buffer[entry->seq_num() % new_size] = std::move(entry);
+      new_buffer[Index(entry->seq_num(), new_size)] = std::move(entry);
     }
   }
   buffer_ = std::move(new_buffer);
@@ -216,11 +215,9 @@ bool PacketBuffer::ExpandBufferSize() {
   return true;
 }
 
-bool PacketBuffer::PotentialNewFrame(uint16_t seq_num) const {
-  size_t index = seq_num % buffer_.size();
-  int prev_index = index > 0 ? index - 1 : buffer_.size() - 1;
-  const auto& entry = buffer_[index];
-  const auto& prev_entry = buffer_[prev_index];
+bool PacketBuffer::PotentialNewFrame(int64_t seq_num) const {
+  const auto& entry = buffer_[Index(seq_num)];
+  const auto& prev_entry = buffer_[Index(seq_num - 1)];
 
   if (entry == nullptr)
     return false;
@@ -230,7 +227,7 @@ bool PacketBuffer::PotentialNewFrame(uint16_t seq_num) const {
     return true;
   if (prev_entry == nullptr)
     return false;
-  if (prev_entry->seq_num() != static_cast<uint16_t>(entry->seq_num() - 1))
+  if (prev_entry->seq_num() != entry->seq_num() - 1)
     return false;
   if (prev_entry->timestamp != entry->timestamp)
     return false;
@@ -241,13 +238,13 @@ bool PacketBuffer::PotentialNewFrame(uint16_t seq_num) const {
 }
 
 std::vector<std::unique_ptr<PacketBuffer::Packet>> PacketBuffer::FindFrames(
-    uint16_t seq_num) {
+    int64_t seq_num) {
   std::vector<std::unique_ptr<PacketBuffer::Packet>> found_frames;
-  auto start = seq_num;
+  int64_t start = seq_num;
 
   for (size_t i = 0; i < buffer_.size(); ++i) {
-    if (received_padding_.find(seq_num) != received_padding_.end()) {
-      seq_num += 1;
+    if (received_padding_.contains(seq_num)) {
+      ++seq_num;
       continue;
     }
 
@@ -255,13 +252,13 @@ std::vector<std::unique_ptr<PacketBuffer::Packet>> PacketBuffer::FindFrames(
       break;
     }
 
-    size_t index = seq_num % buffer_.size();
+    size_t index = Index(seq_num);
     buffer_[index]->continuous = true;
 
     // If all packets of the frame is continuous, find the first packet of the
     // frame and add all packets of the frame to the returned packets.
     if (buffer_[index]->is_last_packet_in_frame()) {
-      uint16_t start_seq_num = seq_num;
+      int64_t start_seq_num = seq_num;
 
       // Find the start index by searching backward until the packet with
       // the `frame_begin` flag is set.
@@ -343,6 +340,7 @@ std::vector<std::unique_ptr<PacketBuffer::Packet>> PacketBuffer::FindFrames(
         // See: https://bugs.chromium.org/p/webrtc/issues/detail?id=7106
         if (is_h264_descriptor &&
             (buffer_[start_index] == nullptr ||
+             buffer_[start_index]->seq_num() != start_seq_num - 1 ||
              buffer_[start_index]->timestamp != frame_timestamp)) {
           break;
         }
@@ -365,7 +363,7 @@ std::vector<std::unique_ptr<PacketBuffer::Packet>> PacketBuffer::FindFrames(
         // Now that we have decided whether to treat this frame as a key frame
         // or delta frame in the frame buffer, we update the field that
         // determines if the RtpFrameObject is a key frame or delta frame.
-        const size_t first_packet_index = start_seq_num % buffer_.size();
+        const size_t first_packet_index = Index(start_seq_num);
         if (is_h264_keyframe) {
           buffer_[first_packet_index]->video_header.frame_type =
               VideoFrameType::kVideoFrameKey;
@@ -389,12 +387,11 @@ std::vector<std::unique_ptr<PacketBuffer::Packet>> PacketBuffer::FindFrames(
       }
 
       if (is_h264_descriptor || full_frame_found) {
-        const uint16_t end_seq_num = seq_num + 1;
-        // Use uint16_t type to handle sequence number wrap around case.
-        uint16_t num_packets = end_seq_num - start_seq_num;
+        const int64_t end_seq_number = seq_num + 1;
+        int64_t num_packets = end_seq_number - start_seq_num;
         found_frames.reserve(found_frames.size() + num_packets);
-        for (uint16_t j = start_seq_num; j != end_seq_num; ++j) {
-          std::unique_ptr<Packet>& packet = buffer_[j % buffer_.size()];
+        for (int64_t j = start_seq_num; j != end_seq_number; ++j) {
+          std::unique_ptr<Packet>& packet = buffer_[Index(j)];
           RTC_DCHECK(packet);
           RTC_DCHECK_EQ(j, packet->seq_num());
           // Ensure frame boundary flags are properly set.
@@ -414,23 +411,24 @@ std::vector<std::unique_ptr<PacketBuffer::Packet>> PacketBuffer::FindFrames(
   return found_frames;
 }
 
-void PacketBuffer::UpdateMissingPackets(uint16_t seq_num) {
+void PacketBuffer::UpdateMissingPackets(int64_t seq_num) {
   if (!newest_inserted_seq_num_)
     newest_inserted_seq_num_ = seq_num;
 
   const int kMaxPaddingAge = 1000;
-  if (AheadOf(seq_num, *newest_inserted_seq_num_)) {
-    uint16_t old_seq_num = seq_num - kMaxPaddingAge;
-    auto erase_to = missing_packets_.lower_bound(old_seq_num);
+  if (seq_num > *newest_inserted_seq_num_) {
+    int64_t old_seq_number = seq_num - kMaxPaddingAge;
+    auto erase_to = missing_packets_.lower_bound(old_seq_number);
     missing_packets_.erase(missing_packets_.begin(), erase_to);
 
     // Guard against inserting a large amount of missing packets if there is a
     // jump in the sequence number.
-    if (AheadOf(old_seq_num, *newest_inserted_seq_num_))
-      *newest_inserted_seq_num_ = old_seq_num;
+    if (old_seq_number > *newest_inserted_seq_num_) {
+      *newest_inserted_seq_num_ = old_seq_number;
+    }
 
     ++*newest_inserted_seq_num_;
-    while (AheadOf(seq_num, *newest_inserted_seq_num_)) {
+    while (seq_num > *newest_inserted_seq_num_) {
       missing_packets_.insert(*newest_inserted_seq_num_);
       ++*newest_inserted_seq_num_;
     }
diff --git a/modules/video_coding/packet_buffer.h b/modules/video_coding/packet_buffer.h
index e205032fa6..9a3e66d761 100644
--- a/modules/video_coding/packet_buffer.h
+++ b/modules/video_coding/packet_buffer.h
@@ -11,6 +11,7 @@
 #ifndef MODULES_VIDEO_CODING_PACKET_BUFFER_H_
 #define MODULES_VIDEO_CODING_PACKET_BUFFER_H_
 
+#include <bit>
 #include <cstddef>
 #include <cstdint>
 #include <memory>
@@ -22,8 +23,8 @@
 #include "api/video/video_codec_type.h"
 #include "modules/rtp_rtcp/source/rtp_packet_received.h"
 #include "modules/rtp_rtcp/source/rtp_video_header.h"
+#include "rtc_base/checks.h"
 #include "rtc_base/copy_on_write_buffer.h"
-#include "rtc_base/numerics/sequence_number_util.h"
 
 namespace webrtc {
 namespace video_coding {
@@ -51,7 +52,7 @@ class PacketBuffer {
     bool is_last_packet_in_frame() const {
       return video_header.is_last_packet_in_frame;
     }
-    uint16_t seq_num() const { return static_cast<uint16_t>(sequence_number); }
+    int64_t seq_num() const { return sequence_number; }
 
     // If all its previous packets have been inserted into the packet buffer.
     // Set and used internally by the PacketBuffer.
@@ -78,11 +79,11 @@ class PacketBuffer {
 
   ABSL_MUST_USE_RESULT InsertResult
   InsertPacket(std::unique_ptr<Packet> packet);
-  ABSL_MUST_USE_RESULT InsertResult InsertPadding(uint16_t seq_num);
+  ABSL_MUST_USE_RESULT InsertResult InsertPadding(int64_t seq_num);
 
   // Clear all packets older than |seq_num|. Returns the number of packets
   // cleared.
-  uint32_t ClearTo(uint16_t seq_num);
+  uint32_t ClearTo(int64_t seq_num);
   void Clear();
 
   void ForceSpsPpsIdrIsH264Keyframe();
@@ -95,19 +96,27 @@ class PacketBuffer {
   bool ExpandBufferSize();
 
   // Test if all previous packets has arrived for the given sequence number.
-  bool PotentialNewFrame(uint16_t seq_num) const;
+  bool PotentialNewFrame(int64_t seq_num) const;
 
   // Test if all packets of a frame has arrived, and if so, returns packets to
   // create frames.
-  std::vector<std::unique_ptr<Packet>> FindFrames(uint16_t seq_num);
+  std::vector<std::unique_ptr<Packet>> FindFrames(int64_t seq_num);
 
-  void UpdateMissingPackets(uint16_t seq_num);
+  void UpdateMissingPackets(int64_t seq_num);
+
+  static size_t Index(int64_t seq_num, size_t buffer_size) {
+    RTC_DCHECK(std::has_single_bit(buffer_size));
+    size_t mask = buffer_size - 1;
+    return seq_num & mask;
+  }
+
+  size_t Index(int64_t seq_num) const { return Index(seq_num, buffer_.size()); }
 
   // buffer_.size() and max_size_ must always be a power of two.
   const size_t max_size_;
 
   // The fist sequence number currently in the buffer.
-  uint16_t first_seq_num_;
+  int64_t first_seq_num_;
 
   // If the packet buffer has received its first packet.
   bool first_packet_received_;
@@ -119,10 +128,9 @@ class PacketBuffer {
   // determine continuity between them.
   std::vector<std::unique_ptr<Packet>> buffer_;
 
-  std::optional<uint16_t> newest_inserted_seq_num_;
-  std::set<uint16_t, DescendingSeqNumComp<uint16_t>> missing_packets_;
-
-  std::set<uint16_t, DescendingSeqNumComp<uint16_t>> received_padding_;
+  std::optional<int64_t> newest_inserted_seq_num_;
+  std::set<int64_t> missing_packets_;
+  std::set<int64_t> received_padding_;
 
   // Indicates if we should require SPS, PPS, and IDR for a particular
   // RTP timestamp to treat the corresponding frame as a keyframe.
diff --git a/modules/video_coding/packet_buffer_unittest.cc b/modules/video_coding/packet_buffer_unittest.cc
index 29f13e4f3e..86e337783e 100644
--- a/modules/video_coding/packet_buffer_unittest.cc
+++ b/modules/video_coding/packet_buffer_unittest.cc
@@ -54,7 +54,7 @@ std::vector<uint16_t> StartSeqNums(
   for (const auto& packet : packets) {
     EXPECT_EQ(frame_boundary, packet->is_first_packet_in_frame());
     if (packet->is_first_packet_in_frame()) {
-      result.push_back(packet->seq_num());
+      result.push_back(static_cast<uint16_t>(packet->seq_num()));
     }
     frame_boundary = packet->is_last_packet_in_frame();
   }
@@ -709,6 +709,14 @@ TEST_P(PacketBufferH264ParameterizedTest, InsertTooOldPackets) {
   InsertH264(4662, kKeyFrame, kFirst, kLast, 1000);
 }
 
+TEST_P(PacketBufferH264ParameterizedTest, InsertMisOrderedPackets) {
+  InsertH264(4660, kKeyFrame, kFirst, kNotLast, 1000);
+  // packet (4661 + kStartSize) can use the same buffer slot as packet 4661,
+  // PacketBuffer need to be careful to detect they are not the same packet.
+  InsertH264(4661 + kStartSize, kDeltaFrame, kFirst, kNotLast, 1000);
+  InsertH264(4662, kKeyFrame, kFirst, kLast, 1000);
+}
+
 TEST_P(PacketBufferH264ParameterizedTest, ClearMissingPacketsOnKeyframe) {
   InsertH264(0, kKeyFrame, kFirst, kLast, 1000);
   InsertH264(2, kKeyFrame, kFirst, kLast, 3000);
diff --git a/video/rtp_video_stream_receiver2.cc b/video/rtp_video_stream_receiver2.cc
index 450e958fbc..419a305077 100644
--- a/video/rtp_video_stream_receiver2.cc
+++ b/video/rtp_video_stream_receiver2.cc
@@ -899,13 +899,13 @@ void RtpVideoStreamReceiver2::OnInsertedPacket(
 
       const video_coding::PacketBuffer::Packet& last_packet = *packet;
       OnAssembledFrame(std::make_unique<RtpFrameObject>(
-          first_packet->seq_num(),  //
-          last_packet.seq_num(),    //
-          last_packet.marker_bit,   //
-          max_nack_count,           //
-          min_recv_time,            //
-          max_recv_time,            //
-          first_packet->timestamp,  //
+          static_cast<uint16_t>(first_packet->seq_num()),  //
+          static_cast<uint16_t>(last_packet.seq_num()),    //
+          last_packet.marker_bit,                          //
+          max_nack_count,                                  //
+          min_recv_time,                                   //
+          max_recv_time,                                   //
+          first_packet->timestamp,                         //
           absolute_capture_time_ms.has_value()
               ? *absolute_capture_time_ms
               : ntp_estimator_.Estimate(first_packet->timestamp),  //
@@ -1194,8 +1194,9 @@ void RtpVideoStreamReceiver2::ReceivePacket(const RtpPacketReceived& packet) {
     // Padding or keep-alive packet.
     // TODO(nisse): Could drop empty packets earlier, but need to figure out how
     // they should be counted in stats.
-    NotifyReceiverOfEmptyPacket(packet.SequenceNumber(),
-                                GetCodecFromPayloadType(packet.PayloadType()));
+    NotifyReceiverOfEmptyPacket(
+        rtp_seq_num_unwrapper_.Unwrap(packet.SequenceNumber()),
+        GetCodecFromPayloadType(packet.PayloadType()));
     return;
   }
   if (packet.PayloadType() == red_payload_type_) {
@@ -1267,8 +1268,9 @@ void RtpVideoStreamReceiver2::ParseAndHandleEncapsulatingHeader(
   if (packet.payload()[0] == ulpfec_receiver_->ulpfec_payload_type()) {
     // Notify video_receiver about received FEC packets to avoid NACKing these
     // packets.
-    NotifyReceiverOfEmptyPacket(packet.SequenceNumber(),
-                                GetCodecFromPayloadType(packet.PayloadType()));
+    NotifyReceiverOfEmptyPacket(
+        rtp_seq_num_unwrapper_.Unwrap(packet.SequenceNumber()),
+        GetCodecFromPayloadType(packet.PayloadType()));
   }
   if (ulpfec_receiver_->AddReceivedRedPacket(packet)) {
     ulpfec_receiver_->ProcessReceivedFec();
@@ -1279,8 +1281,9 @@ void RtpVideoStreamReceiver2::ParseAndHandleEncapsulatingHeader(
 // RtpFrameReferenceFinder will need to know about padding to
 // correctly calculate frame references.
 void RtpVideoStreamReceiver2::NotifyReceiverOfEmptyPacket(
-    uint16_t seq_num,
+    int64_t seq_number,
     std::optional<VideoCodecType> codec) {
+  uint16_t seq_num = static_cast<uint16_t>(seq_number);
   RTC_DCHECK_RUN_ON(&packet_sequence_checker_);
   RTC_DCHECK_RUN_ON(&worker_task_checker_);
 
@@ -1289,7 +1292,7 @@ void RtpVideoStreamReceiver2::NotifyReceiverOfEmptyPacket(
   if (h26x_packet_buffer_ && UseH26xPacketBuffer(codec)) {
     OnInsertedPacket(h26x_packet_buffer_->InsertPadding(seq_num));
   } else {
-    OnInsertedPacket(packet_buffer_.InsertPadding(seq_num));
+    OnInsertedPacket(packet_buffer_.InsertPadding(seq_number));
   }
   if (nack_module_) {
     nack_module_->OnReceivedPacket(seq_num, /*is_recovered=*/false);
@@ -1367,7 +1370,7 @@ void RtpVideoStreamReceiver2::FrameDecoded(int64_t picture_id) {
     int64_t unwrapped_rtp_seq_num = rtp_seq_num_unwrapper_.Unwrap(seq_num);
     packet_infos_.erase(packet_infos_.begin(),
                         packet_infos_.upper_bound(unwrapped_rtp_seq_num));
-    uint32_t num_packets_cleared = packet_buffer_.ClearTo(seq_num);
+    uint32_t num_packets_cleared = packet_buffer_.ClearTo(unwrapped_rtp_seq_num);
     if (num_packets_cleared > 0) {
       TRACE_EVENT2("webrtc",
                    "RtpVideoStreamReceiver2::FrameDecoded Cleared Old Packets",
diff --git a/video/rtp_video_stream_receiver2.h b/video/rtp_video_stream_receiver2.h
index 42f44aa6f8..a5b745bd11 100644
--- a/video/rtp_video_stream_receiver2.h
+++ b/video/rtp_video_stream_receiver2.h
@@ -324,7 +324,7 @@ class RtpVideoStreamReceiver2 : public LossNotificationSender,
   // This function assumes that it's being called from only one thread.
   void ParseAndHandleEncapsulatingHeader(const RtpPacketReceived& packet)
       RTC_RUN_ON(packet_sequence_checker_);
-  void NotifyReceiverOfEmptyPacket(uint16_t seq_num,
+  void NotifyReceiverOfEmptyPacket(int64_t seq_number,
                                    std::optional<VideoCodecType> codec)
       RTC_RUN_ON(packet_sequence_checker_);
   bool IsRedEnabled() const;
